oeps/oep-0049-django-app-patterns.rst

```diff
@@ -0,0 +1,165 @@
+
+
+Not exposing an app's data structures can be tricky because it's very easy to expose them without meaning to. Therefore there are a couple common strategies we employ.
+
+1. When importing internal app code to be used in the ``api.py`` file, prefix it with an underscore so it's clear it's for internal use only.
+
+2. Create a ``data.py`` file to house simple data objects that can be passed from your app's function to the calling app. By creating these objects, we can avoid both passing Django model objects or querysets directly and having to serialize data. Other apps may import data classes from ``data.py`` in additional to functionality from ``api.py``. See data.py_ for more details.
+
+For example:
+
+.. code-block:: python
+
+    from django.conf.settings import UNSUPPORTED_PROGRAM_UUIDS
+
+    from .data import ProgramData
+    from .models_api import get_programs as _get_programs
+
+    def get_supported_programs():
+        """
+        Gets all programs that aren't in UNSUPPORTED_PROGRAM_UUIDS settings
+        """
+        all_programs = _get_programs() # returns queryset
+        supported_programs [
+          ProgramData(
+            uuid=program.uuid,
+            title=program.title,
+            status=program.status
+          )
+          for program in all_programs
+          if program.uuid not in UNSUPPORTED_PROGRAM_UUIDS
+        ]
+        return supported_programs
+
+.. _data.py:
+data.py
++++++++
+This file should include the public data structures for the app that can be passed between apps without exposing internal features. These should be used instead of sending Django model objects or querysets to apps that call the functions in ``api.py``. This file should not import anything other than stdlib modules, so that it may be imported by any other app without issue. These data objects should be simple objects with all business logic handled by ``api.py``. They may however perform simple validation, as long as it is self-contained (doesn't reach out to database, network, or any code outside of the class)
+
+For example:
+
+.. code-block:: python
+
+
+    from dataclasses import dataclass
+    from enum import Enum
+
+    def ProgramStatus(Enum):
+        ACTIVE = "active"
+        RETIRED = "retired"
+
+    @dataclass
+    class ProgramData:
+        uuid: str
+        title: str
+
+.. _models_api.py:
+models_api.py
++++++++++++++
+.. Reviewer NOTE: I don't feel like I entirely understand models_api.py. My main issue is I don't understand the benefit of the additional layer versus api.py just calling ``Thing.objects.get()``.
+
+
+.. _rest_api:
+rest_api/
++++++++++
+Each app should self-contain it's related REST API, meaning it should live within the app and not in a separate "api" app where it would mingle with other apps' REST API code. Since the Python API is using ``api.py``, the REST API code (often Django Rest Framework) should live in a ``rest_api`` folder.
```

Comment:
I'm not sure if it's worthwhile to deviate from having these in `views.py`. It's where people will expect to look for views code, including DRF views. And we're not really building any other kind of view.

Also, I suspect if we really go all-in on this pattern, we'll find that REST APIs do not align 1:1 with apps and their internal APIs. For instance, with `learning_sequences`, I hope that eventually many other REST APIs query it to get course outline and sequence metadata, marked up with whatever additional information makes sense for them. It's kind of like Netflix's backend-for-frontend, where instead of having a bunch of backend services that get aggregated into different REST APIs, we get a bunch of small in-process API calls that get composed.

Reply:
I guess I haven't been paying enough attention–I knew the `rest_api` practice existed, but I didn't realize it was that widespread. FWIW, it's not something I feel strongly about, particularly since there's a `views.py` in that directory anyhow.

Reply:
Though I would point out that we're not really making any other kind of view these days.